### PR TITLE
Add maSigPro wrapper

### PR DESCRIPTION
**Pull request by blankclemens**
This adds a wrapper for [maSigPro](https://bioconductor.org/packages/release/bioc/html/maSigPro.html).

Besides using native edesign and data files, the wrapper allows to build an experiment design from separate count files ( e.g. from HTSeq-count) and can be used in such workflows.

### CODE REVIEW

File: tools/masigpro/masigpro.xml [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+                    <(cut -f1 $file)
+                    #set $start = False
+                #end if
+                #set $header += ' "' + $file.name + '"'
+                <(cut -f2 $file)
+            #end for
+        #end for
+        > data && sed -i '1i$header' data &&
+        #if $source.enable_output:
+            cp data $data_out && cp $design_matrix $edesign_out &&
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: mv maybe?

File: tools/masigpro/masigpro.xml [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        #if str($pdf.seeGenes.clustering.clusterMethod) == "kmeans":
+            --iter_max $pdf.seeGenes.clustering.iterMax
+        #end if
+        --color_mode $pdf.seeGenes.colorMode
+        --show_fit $pdf.seeGenes.showFit
+        --show_lines $pdf.seeGenes.showLines
+        --cexlab $pdf.seeGenes.cexlab
+        --legend $pdf.seeGenes.legend
+    #end if
+    && mv Results.pdf $pdf_out
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: You could avoid this and just use `from_working_dir`

File: tools/masigpro/masigpro.xml [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+            <option value="two.ways.forward">two.ways.forward</option>
+            </param>
+            <param type="boolean" name="nvar_correction" label="nvar correction" truevalue="TRUE" falsevalue="FALSE" checked="false"
+                help="When nvar.correction is TRUE the given significance
+                    level is corrected by the number of variables in the model.">
+            <option selected="True" value="FALSE">False</option>
+            <option value="TRUE">True</option>
+            </param>
+            <param name="alfa" type="float" value="0.05" label="alfa" help="Significance level used for variable selection in the stepwise regression"
+            />
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: This looks wired.

File: tools/masigpro/masigpro.xml [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+                    threshold alfa.">
+            <option selected="True" value="backward">backward</option>
+            <option value="forward">forward</option>
+            <option value="two.ways.backward">two.ways.backward</option>
+            <option value="two.ways.forward">two.ways.forward</option>
+            </param>
+            <param type="boolean" name="nvar_correction" label="nvar correction" truevalue="TRUE" falsevalue="FALSE" checked="false"
+                help="When nvar.correction is TRUE the given significance
+                    level is corrected by the number of variables in the model.">
+            <option selected="True" value="FALSE">False</option>
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: indentation

File: tools/masigpro/masigpro.xml [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+            <param name="step_method" type="select" label="Step regression"
+                help="The step regression can be ‘backward’ or ‘forward’ indicating
+                    whether the step procedure starts from the model with all or none
+                    variables. With the ‘two.ways.backward’ or ‘two.ways.forward’
+                    options the variables are both allowed to get in and out. At each
+                    step the p-value of each variable is computed and variables get
+                    in/out the model when this p-value is lower or higher than given
+                    threshold alfa.">
+            <option selected="True" value="backward">backward</option>
+            <option value="forward">forward</option>
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: Indentation

File: tools/masigpro/masigpro.xml [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+#end if
+</configfile>
+    </configfiles>
+    <inputs>
+        <conditional name="source">
+            <param label="Choose data source" name="source_selector"
+                help="Choose if you want to provide seperate count files (e.g. from HTSeq-count or feature-seq) 
+                    and define your experiment design matrix here, or if you have maSigPro edesign and data input files already."
+                type="select">
+            <option value="defaults">Use maSigPro edesign and data files</option>
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: indentation

File: tools/masigpro/masigpro.xml [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+                        <repeat name="rep_repl" title="Replicate groups" min="1" default="1">
+                            <param name="files" type="data" format="tabular, txt" multiple="true" label="Counts file(s) that are replicates" />
+                        </repeat>
+                    </when>
+                </conditional>
+                <repeat name="rep_groups" title="Experimental groups" min="1" default="1">
+                    <param name="name" type="text" value="Group title" label="Specify the name of this experimental group"
+                        help="Use a single name without spaces or special characters">
+                    </param>
+                    <param name="files" type="data" format="tabular, txt" multiple="true"
```
Comment (bgruening) [inline tools/masigpro/masigpro.xml]: txt, is really allowed?

**Comment by blankclemens** [comment]
Thank a lot for the comments, they should be fixed now.

**Comment by bgruening** [comment]
Two last questions. Have you considered to use an other configfile for the R code and avoid passing around arguments?

@galaxyproject/iuc how do you feel about sarcasm in wrappers, like this one? https://github.com/galaxyproject/tools-iuc/pull/1303/files#diff-85ad6ddb8369d0149c5eeb266cdd1dc7R167

I have mixed feelings here and after sleeping I think I tend to remove it, especially as it is the default option. Having a conditional with only one param is useless and the wording might be offending to a few (like the FSM discussion we once had). A warning might be more appropriate?

**Comment by blankclemens** [comment]
I don't see a a huge value there for normal arguments - letting you have a correct R file on the contrary allows to edit and test the R file properly. Theoretically the R file can stand on it's own, while the creation of the experiment matrix is directly bound to the wrapper.

It could be nice however to be able to specify an external configfile, since it can get quite uncomfortable to write inside the wrapper.

Relating the sarcasm: I don't mind offending, but I also don't mind removing it. I can't however see how this conditional is useless - it's a repeat parameter after all.

**Comment by bgruening** [comment]
>  I can't however see how this conditional is useless - it's a repeat parameter after all.

You can simply remove the conditional and start the repeat with min=0, aren't you?

**Comment by blankclemens** [comment]
Haha, for some reason I haven't thought about 0 being a possibility here. Well played, I'll look into that tomorrow.

**Comment by bgruening** [comment]
Cool, thanks @blankclemens!

**Pull request merged**